pebble_merger.go

```diff
@@ -1,58 +1,126 @@
 package dhstore
 
 import (
+	"bytes"
 	"io"
 
 	"github.com/cockroachdb/pebble"
 )
 
-const deletableDefaultMergerName = "dhstore.v1.deletable_default_merger"
+const valueKeysMergerName = "dhstore.v1.valueKeysMerger"
 
 var (
-	_ pebble.ValueMerger          = (*deletableDefaultMerger)(nil)
-	_ pebble.DeletableValueMerger = (*deletableDefaultMerger)(nil)
+	_ pebble.ValueMerger          = (*valueKeysValueMerger)(nil)
+	_ pebble.DeletableValueMerger = (*valueKeysValueMerger)(nil)
 )
 
-type deletableDefaultMerger struct {
-	def pebble.ValueMerger
+type valueKeysValueMerger struct {
+	merges             []EncryptedValueKey
+	reverse            bool
+	marshalledSizeHint int // Used as a hint to grow the buffer size during marshalling.
+	s                  *PebbleDHStore
 }
 
-func newDeletableDefaultMerger() *pebble.Merger {
+func (s *PebbleDHStore) newValueKeysMerger() *pebble.Merger {
 	return &pebble.Merger{
 		Merge: func(k, value []byte) (pebble.ValueMerger, error) {
 			// Fall back on default merger if the key is not of type multihash, i.e. the only key
 			// type that corresponds to value-keys.
 			switch keyPrefix(k[0]) {
 			case multihashKeyPrefix:
-				merge, err := pebble.DefaultMerger.Merge(k, value)
-				if err != nil {
-					return nil, err
-				}
-				v := &deletableDefaultMerger{
-					def: merge,
-				}
-				return v, nil
+				v := &valueKeysValueMerger{s: s}
+				return v, v.MergeNewer(value)
 			default:
 				return pebble.DefaultMerger.Merge(k, value)
 			}
 		},
-		Name: deletableDefaultMergerName,
+		Name: valueKeysMergerName,
 	}
 }
 
-func (v *deletableDefaultMerger) MergeNewer(value []byte) error {
-	return v.def.MergeNewer(value)
+func (v *valueKeysValueMerger) MergeNewer(value []byte) error {
+	if len(value) == 0 {
+		return nil
+	}
+
+	evks, err := v.s.unmarshalEncryptedIndexKeys(value)
+	if err != nil {
+		return err
+	}
+
+	v.merges = maybeGrow(v.merges, len(evks))
+
+	if len(v.merges) == 0 {
+		// Optimise for the case where there are no merges.
+		v.merges = append(v.merges, evks...)
+		v.marshalledSizeHint += len(value)
+	} else {
+		for _, evk := range evks {
+			if !v.exists(evk) {
+				v.merges = append(v.merges, evk)
+				v.marshalledSizeHint += len(evk)
+			}
+		}
+	}
+
+	return nil
 }
 
-func (v *deletableDefaultMerger) MergeOlder(value []byte) error {
-	return v.def.MergeOlder(value)
+func (v *valueKeysValueMerger) MergeOlder(value []byte) error {
+	v.reverse = true
+	return v.MergeNewer(value)
 }
 
-func (v *deletableDefaultMerger) Finish(includesBase bool) ([]byte, io.Closer, error) {
-	return v.def.Finish(includesBase)
+func (v *valueKeysValueMerger) Finish(_ bool) ([]byte, io.Closer, error) {
+	if len(v.merges) == 0 {
+		return nil, nil, nil
+	}
+	if v.reverse {
+		for one, other := 0, len(v.merges)-1; one < other; one, other = one+1, other-1 {
+			v.merges[one], v.merges[other] = v.merges[other], v.merges[one]
```

Comment:
Can you please explain logic behind this loop?

Reply:
See `pebble.ValueMerger` [docs](https://pkg.go.dev/github.com/cockroachdb/pebble@v0.0.0-20230113195140-25d2b70d1aeb/internal/base#ValueMerger) on the requirements around transitivity.

Depending on weather MergerNewer or MergerOlder is called first we need to reverse the order to respect transitive property. There is a unit test that asserts this property too.